File: frontend/types/appointment.ts
/**
 * appointment.ts — Appointment, chat, file and participant types.
 *
 * 1:1 mapping with backend/app/schemas/student.py where available.
 * Chat / file / participant / detail responses are not yet shipped by the
 * backend (ROADMAP 3.6) and are marked with explicit TODOs.
 */

import type {
  AppointmentStatus,
  ConsultationType,
  IsoDateTime,
  ParticipantStatus,
  TopicCategory,
  Uuid,
} from "./common"

// ── Request/response: booking flow ───────────────────────────────────────────
// Source: backend/app/schemas/student.py

export interface AppointmentCreateRequest {
  slot_id: Uuid
  topic_category: TopicCategory
  description: string
  subject_id?: Uuid | null
}

export interface AppointmentResponse {
  id: Uuid
  slot_id: Uuid
  professor_id: Uuid
  lead_student_id: Uuid
  subject_id: Uuid | null
  topic_category: TopicCategory
  description: string
  status: AppointmentStatus
  consultation_type: ConsultationType
  slot_datetime: IsoDateTime
  created_at: IsoDateTime
}

export interface AppointmentCancelResponse {
  id: Uuid
  status: AppointmentStatus
}

// ── Appointment detail + chat + files + participants ─────────────────────────
// TODO: sync with backend when schema is implemented (ROADMAP 3.6 introduces
//       /appointments/{id} detail + chat/messages + file + participant endpoints).

export interface AppointmentDetailResponse extends AppointmentResponse {
  /** Flag from models/appointment.py: true when slot.max_students > 1. */
  is_group: boolean
  /** Populated when professor delegates approval to an assistant. */
  delegated_to: Uuid | null
  /** Populated when professor rejects the appointment. */
  rejection_reason: string | null
  /** Count of chat messages — drives "X/20" counter. */
  chat_message_count: number
  /** Count of uploaded files. */
  file_count: number
}

export interface ChatMessageResponse {
  id: Uuid
  appointment_id: Uuid
  sender_id: Uuid
  content: string
  created_at: IsoDateTime
}

export interface ChatMessageCreate {
  content: string
}

export interface FileResponse {
  id: Uuid
  appointment_id: Uuid
  uploaded_by: Uuid
  filename: string
  mime_type: string
  file_size_bytes: number
  created_at: IsoDateTime
  /** Presigned MinIO download URL (time-limited). Populated on GET detail. */
  download_url?: string
}

export interface ParticipantResponse {
  id: Uuid
  appointment_id: Uuid
  student_id: Uuid
  status: ParticipantStatus
  is_lead: boolean
  confirmed_at: IsoDateTime | null
  /** Convenience: student full name (backend may expose via join). */
  student_full_name?: string
}

// ── Waitlist (student) ────────────────────────────────────────────────────────
// Source: backend/app/api/v1/students.py — POST/DELETE /waitlist/{slot_id}
// Backend returns MessageResponse with "Pozicija: N" in the text; we parse
// that on the client if needed. A dedicated response shape is not yet exposed.

export interface WaitlistPosition {
  slot_id: Uuid
  position: number
}

File: backend/app/schemas/appointment.py
"""Pydantic V2 schemas for the appointment detail / chat / files / participants
flow (Faza 3.3).

Each schema is paired 1:1 with the locked frontend contract in
``frontend/types/appointment.ts``. Fields use snake_case on both sides
(see CLAUDE.md §17 + CURRENT_STATE2.md §7).

Note on construction:
    The existing project convention (``backend/app/api/v1/students.py``) builds
    ``AppointmentResponse`` manually because ``slot_datetime`` requires a join
    on ``appointment.slot.slot_datetime``. We follow the same pattern here —
    services hydrate these schemas explicitly rather than relying on
    ``from_attributes`` traversing relationships.
"""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field

from app.models.enums import (
    AppointmentStatus,
    ConsultationType,
    ParticipantStatus,
    TopicCategory,
)


# ── Detail ────────────────────────────────────────────────────────────────────


class AppointmentDetailResponse(BaseModel):
    """Mirrors ``frontend/types/appointment.ts::AppointmentDetailResponse``.

    Flat shape — files / participants / messages live behind separate endpoints
    (``GET /{id}/files``, ``GET /{id}/participants``, ``GET /{id}/messages``).
    Counts here drive UI badges (e.g. "X/20" chat counter, file tab badge).
    """

    id: UUID
    slot_id: UUID
    professor_id: UUID
    lead_student_id: UUID
    subject_id: UUID | None
    topic_category: TopicCategory
    description: str
    status: AppointmentStatus
    consultation_type: ConsultationType
    slot_datetime: datetime
    created_at: datetime
    is_group: bool
    delegated_to: UUID | None
    rejection_reason: str | None
    chat_message_count: int
    file_count: int


# ── Chat (REST polling fallback; WS upgrade in Faza 4.1) ──────────────────────


class ChatMessageResponse(BaseModel):
    """Mirrors ``frontend/types/appointment.ts::ChatMessageResponse``.

    Flat REST shape (sender_id only). The WS chat envelope from
    ``docs/websocket-schema.md §5`` uses a nested ``sender`` object — that is a
    separate contract delivered with Faza 4.1 and intentionally not collapsed
    into this REST schema.
    """

    id: UUID
    appointment_id: UUID
    sender_id: UUID
    content: str
    created_at: datetime


class ChatMessageCreate(BaseModel):
    """Mirrors ``frontend/types/appointment.ts::ChatMessageCreate``.

    ``content`` is bounded by Pydantic; whitespace-only strings are rejected
    in the service layer (after ``.strip()``).
    """

    content: str = Field(min_length=1, max_length=1000)


# ── Files ─────────────────────────────────────────────────────────────────────


class FileResponse(BaseModel):
    """Mirrors ``frontend/types/appointment.ts::FileResponse``.

    ``download_url`` is a presigned MinIO GET URL with a 1h TTL. It is
    populated on list / upload responses and absent on bare DB reads.
    """

    id: UUID
    appointment_id: UUID
    uploaded_by: UUID
    filename: str
    mime_type: str
    file_size_bytes: int
    created_at: datetime
    download_url: str | None = None


# ── Participants (group consultations) ────────────────────────────────────────


class ParticipantResponse(BaseModel):
    """Mirrors ``frontend/types/appointment.ts::ParticipantResponse``.

    ``student_full_name`` is an optional convenience field hydrated via a join
    on ``users`` so the UI can render the participant list without a second
    round-trip.
    """

    id: UUID
    appointment_id: UUID
    student_id: UUID
    status: ParticipantStatus
    is_lead: bool
    confirmed_at: datetime | None
    student_full_name: str | None = None
